### DISCUSSION TITLE
Using Waveshare 4.2 E-Paper Display Module (B) on ESP32

### DISCUSSION THREAD

**Discussion by ReiettoAyanami** [ZinggJM/GxEPD2#71] [Q&A] [answered]
Hi,
I'm new to using E-paper display and so I wanted to ask if it was possible to know if
[this](https://amzn.eu/d/g28lvgq) display was compatible with an esp32 and if it was possible to use it as an image displayer (even one at a time is good).
The official webstore: https://www.waveshare.com/4.2inch-e-paper-module-b.htm
My doubt for using this as an image displayer comes from this line in the official GxEPD2 library repo:
`////#include "bitmaps/Bitmaps3c400x300.h" // 4.2"  b/w/r // not enough code space`
this is on [line 137: GxEPD2/examples/GxEPD2_HelloWorld/GxEPD2_HelloWorld.ino](https://github.com/ZinggJM/GxEPD2/blob/master/examples/GxEPD2_HelloWorld/GxEPD2_HelloWorld.ino)
Even though it should only be a 15KB BMP image, no?
I already know it doesn't support partial refresh, and that's not a big deal, I just want to know
I'm using [this ESP32-CH340 from AZ-Delivery](https://amzn.eu/d/26erQxe).

Yesterday I tried with uncommenting [line 280  in examples/GxEPD2_Example/GxEPD2_display_selection.h](https://github.com/ZinggJM/GxEPD2/blob/master/examples/GxEPD2_Example/GxEPD2_display_selection.h):
`// GxEPD2_3C<GxEPD2_420c_Z21, GxEPD2_420c_Z21::HEIGHT> display(GxEPD2_420c_Z21(/*CS=5*/ SS, /*DC=*/ 17, /*RST=*/ 16, /*BUSY=*/ 4)); // GDEQ042Z21 400x300, UC8276`
(old style)
And also tried uncommenting [line 93 in examples/GxEPD2_Example/GxEPD2_display_selection_new_style.h](https://github.com/ZinggJM/GxEPD2/blob/master/examples/GxEPD2_Example/GxEPD2_display_selection_new_style.h)
`//#define GxEPD2_DRIVER_CLASS GxEPD2_420c_Z21 // GDEQ042Z21  400x300, UC8276, (hidden)`

The wiring i'm using is described at [line 74 in examples/GxEPD2_Example/GxEPD2_wiring_examples.h](https://github.com/ZinggJM/GxEPD2/blob/master/examples/GxEPD2_Example/GxEPD2_wiring_examples.h):
`// BUSY -> 4, RST -> 0, DC -> 2, CS -> SS(5), CLK -> SCK(18), DIN -> MOSI(23), GND -> GND, 3.3V -> 3.3V
// for use with Board: "ESP32 Dev Module":`

I tried this the other day and worked, but after some testing it started displaying noise, I've already checked if it's a wiring issue and I'm pretty confident that it's not the case.

I'm now buying another display to check if it's a hardware-related issue, but for now some help would be really appreciated.
Thanks to anyone answering to this post.
Best Regards.

P.S
I'm using Adafruit_GFX not the GFX class provided by GxEPD2.
I've tried using the "new style" and "old style" for initializing the display, neither work after my second attempt, I've tried using the old style alone, and i can at least "talk" to the display, using the new style alone doesn't even initialize the `display` object for me.

**Comment by G6EJD** [comment]
You need to use the half-buffer variant/option

**Reply by ReiettoAyanami** [reply]
Hi,
thanks for your answer, I guess you're referring to this line?
`GxEPD2_3C<GxEPD2_420c_Z21, GxEPD2_420c_Z21::HEIGHT / 2> display(GxEPD2_420c_Z21(/*CS=15*/ SS, /*DC=4*/ 4, /*RST=2*/ 2, /*BUSY=5*/ 5)); // GDEQ042Z21 400x300, UC8276`
(Line 186 in: [examples/GxEPD2_Example/GxEPD2_display_selection.h](https://github.com/ZinggJM/GxEPD2/blob/master/examples/GxEPD2_Example/GxEPD2_display_selection.h))
But I think I'm wrong, since it's defined for ESP8266. (Correct me if I'm wrong on this).

Or is it this line?
`GxEPD2_3C<GxEPD2_420c_Z21, GxEPD2_420c_Z21::HEIGHT> display(GxEPD2_420c_Z21(/*CS=*/ 15, /*DC=*/ 27, /*RST=*/ 26, /*BUSY=*/ 25)); // GDEQ042Z21 400x300, UC8276`
(Line 392 in:  [examples/GxEPD2_Example/GxEPD2_display_selection.h](https://github.com/ZinggJM/GxEPD2/blob/master/examples/GxEPD2_Example/GxEPD2_display_selection.h))
But, still, it's said it's used for Waveshare ESP32 Driver Board:
`// ***** for mapping of Waveshare ESP32 Driver Board *****`
(Line 327 in:  [examples/GxEPD2_Example/GxEPD2_display_selection.h](https://github.com/ZinggJM/GxEPD2/blob/master/examples/GxEPD2_Example/GxEPD2_display_selection.h))

And I can't really find any half buffer variants for ESP32 definitions...

Thanks for all the support!

**Reply by G6EJD** [reply]
Yes, select your display and half-buffer option, then set the pins the same as those connected to your display SPI bus.

**Reply by ReiettoAyanami** [reply]
So, just to be sure, do I need to uncomment this line:
> `GxEPD2_3C<GxEPD2_420c_Z21, GxEPD2_420c_Z21::HEIGHT / 2> display(GxEPD2_420c_Z21(/*CS=15*/ SS, /*DC=4*/ 4, /*RST=2*/ 2, /*BUSY=5*/ 5)); // GDEQ042Z21 400x300, UC8276`
> (Line 186 in: [examples/GxEPD2_Example/GxEPD2_display_selection.h]>(https://github.com/ZinggJM/GxEPD2/blob/master/examples/GxEPD2_Example/GxEPD2_display_selection.h))

or

> `GxEPD2_3C<GxEPD2_420c_Z21, GxEPD2_420c_Z21::HEIGHT> display(GxEPD2_420c_Z21(/*CS=*/ 15, /*DC=*/ 27, /*RST=*/ 26, /*BUSY=*/ 25)); // GDEQ042Z21 400x300, UC8276`
>(Line 392 in:  [examples/GxEPD2_Example/GxEPD2_display_selection.h](https://github.com/ZinggJM/GxEPD2/blob/master/examples/GxEPD2_Example/GxEPD2_display_selection.h))

this line?

It's still a bit unclear to me.
Thanks!

**Reply by G6EJD** [reply]
Yes the first line if that matches your display type Z21. Then note your display connection and make sure all the pins match the numbers on the display definition line. Also make sure MOSI is 23 and CLK 28, but this depends on your display connection.

**Reply by ReiettoAyanami** [reply]
Ok, 
I Will try using the half-buffer line even tough it's defined for ESP8266 when I get back home.
I was using this PIN configuration as suggested in the wiring_examples file: 
`// BUSY -> 4, RST -> 0, DC -> 2, CS -> SS(5), CLK -> SCK(18), DIN -> MOSI(23), GND -> GND, 3.3V -> 3.3V // for use with Board: "ESP32 Dev Module`, and now I should be using the same but with CLK = 28?

Just one last question: 
Would videos and photos be accepted/useful in this case?

Sorry if I'm so desperate for answers, It's for a gift and I'm trying to deliver it before the 25th.
Thanks again, I will be reaching out later, for now, you've been as helpful as you could.

**Reply by G6EJD** [reply]
There is an ESP32 variant for half-buffer if your display is supported that is. 
It should not be causing any difficulty to setup.

**Reply by ReiettoAyanami** [reply]
Edit: (Not an Hardware-related issue. I'am testing it with a raspberry pi and it works fine. I have been testing it with WaveShare example code. The ESP-32 also works just fine.)


So, I've come back home, and this is what I've got:
```cpp
// GxEPD2_HelloWorld.ino by Jean-Marc Zingg (W/O comments)
#include <GxEPD2_3C.h>
#include <Fonts/FreeMonoBold9pt7b.h>
#include "GxEPD2_display_selection.h"
#include "GxEPD2_display_selection_added.h"

void setup()
{
  display.init(115200, true, 2, false);
  helloWorld();
  display.hibernate();
}

const char HelloWorld[] = "Hello World!";

void helloWorld()
{
  display.setRotation(1);
  display.setFont(&FreeMonoBold9pt7b);
  display.setTextColor(GxEPD_BLACK);
  int16_t tbx, tby; uint16_t tbw, tbh;
  display.getTextBounds(HelloWorld, 0, 0, &tbx, &tby, &tbw, &tbh);
  uint16_t x = ((display.width() - tbw) / 2) - tbx;
  uint16_t y = ((display.height() - tbh) / 2) - tby;
  display.setFullWindow();
  display.firstPage();
  do
  {
    display.fillScreen(GxEPD_WHITE);
    display.setCursor(x, y);
    display.print(HelloWorld);
  }
  while (display.nextPage());
}

void loop() {}
```
The display definition I'm using: 
`GxEPD2_3C<GxEPD2_420c_Z21, GxEPD2_420c_Z21::HEIGHT> display(GxEPD2_420c_Z21(/*CS=5*/ SS, /*DC=*/ 17, /*RST=*/ 16, /*BUSY=*/ 4)); // GDEQ042Z21 400x300, UC8276` (Line 280 in: GxEPD2_display_selection.h)

Wiring:
`BUSY -> 4, RST -> 0, DC -> 2, CS -> SS(5), CLK -> SCK(18), DIN -> MOSI(23), GND -> GND, 3.3V -> 3.3V`

Photos:
![WhatsApp Image 2024-12-10 at 20 03 05_f59e1d63](https://github.com/user-attachments/assets/243af37b-27d8-42c9-b84c-edcefd230b4d)
(The situation with the display)
![WhatsApp Image 2024-12-10 at 20 03 06_eae224b6](https://github.com/user-attachments/assets/c0306682-b44a-484c-b5ae-e77f013d9845)
(Display Wiring)
![WhatsApp Image 2024-12-10 at 20 03 06_8a49c94b](https://github.com/user-attachments/assets/25366d85-237d-4f36-8df4-97f3426b45e1)
(ESP-32 Wiring Side A)
![WhatsApp Image 2024-12-10 at 20 03 07_238353c7](https://github.com/user-attachments/assets/dc90e437-f8ff-4005-96d1-ebf32d60c5cd)
(ESP-32 Wiring Side B)
https://github.com/user-attachments/assets/6df44e9c-f7e5-436f-8309-702f40bf7aff
(What running the code looks like)

**Reply by G6EJD** [reply]
The wiring is correct, the display is working, but this very much looks like you have the wrong display object. 

Have you tried the GxEPD2 example snd working your way through the 4.2" variants? That's the best way forward.

Where did you get the display from and have you been able to identify the display driver fitted?

**Reply by ReiettoAyanami** [reply]
I've got the display from Waveshare's amazon marketplace, The link is in the first message (https://amzn.eu/d/g28lvgq), I'm now trying what you are suggesting, I'll tell you how it goes.

**Reply by G6EJD** [reply]
So you have 3-choices:

//GxEPD2_3C<GxEPD2_420c, GxEPD2_420c::HEIGHT / 2> display(GxEPD2_420c(/*CS=15*/ SS, /*DC=4*/ 4, /*RST=2*/ 2, /*BUSY=5*/ 5)); // GDEW042Z15 400x300, UC8176 (IL0398)
//GxEPD2_3C<GxEPD2_420c_Z21, GxEPD2_420c_Z21::HEIGHT / 2> display(GxEPD2_420c_Z21(/*CS=15*/ SS, /*DC=4*/ 4, /*RST=2*/ 2, /*BUSY=5*/ 5)); // GDEQ042Z21 400x300, UC8276
//GxEPD2_3C<GxEPD2_420c_GDEY042Z98, GxEPD2_420c_GDEY042Z98::HEIGHT> display(GxEPD2_420c_GDEY042Z98(/*CS=15*/ SS, /*DC=4*/ 4, /*RST=2*/ 2, /*BUSY=5*/ 5)); // GDEY042Z98 400x300, SSD1683 (no inking)

**Reply by G6EJD** [reply]
You'll need about 45K of free ram to build a 3-colour image, a lot will depend on your final source code s d how much ram is consumed by variables

**Reply by ReiettoAyanami** [reply]
Hi, sorry if I'm taking long, I had to deal with some work related stuff, and i also forgot that you told me to use the half sized buffer like an idiot so i was basically doing no progress.

**Reply by ReiettoAyanami** [reply]
Ok, I've tried them all, they don't even light up the display apparently, the most common serial output after booting is "Busy Timeout!".

**Reply by G6EJD** [reply]
OK, assuming you checked/amended the PIN numbers for each object changed. Both half and full buffer variants should work for that size screen and the GxEPD2 Hello World should be your test example.
Did you uncomment the correct driver object line and change the pin definition?
It is possible the display is faulty or the driver chip used is not yet supported by GxEPD2

**Reply by ReiettoAyanami** [reply]
I eventually bought one of the same display but just black and white, this one: https://www.waveshare.com/4.2inch-e-paper-module.htm
it's the same display, and I'm pretty sure the display is supported by the driver, so I can debug with it.
I will reach out tomorrow when it arrives and let you know if it works, unfortunately tonight I can't really do anything with the display.

**Comment by ZinggJM** [comment, marked as answer]
- you can select the display either in [GxEPD2_display_selection.h](https://github.com/ZinggJM/GxEPD2/blob/master/examples/GxEPD2_Example/GxEPD2_display_selection.h) or in [GxEPD2_display_selection_new_style.h](https://github.com/ZinggJM/GxEPD2/blob/master/examples/GxEPD2_Example/GxEPD2_display_selection_new_style.h), but not in both at the same time.
- you need to make sure that the constructor parameters correspond to your wiring, or your wiring corresponds to the constructor parameters.
- line 74 of [GxEPD2_wiring_examples.h](https://github.com/ZinggJM/GxEPD2/blob/master/examples/GxEPD2_Example/GxEPD2_wiring_examples.h) is for the wiring of the 24pin FPC connector of a TTGO T8 ESP32-WROVER. The corresponding constructor line is line 221 of [GxEPD2_display_selection_new_style.h](https://github.com/ZinggJM/GxEPD2/blob/master/examples/GxEPD2_Example/GxEPD2_display_selection_new_style.h).

You might test your display with the demo code of the Waveshare library https://github.com/waveshareteam/e-Paper. If the demo for your display is newer than the driver class in GxEPD2, then it may not be supported by GxEPD2, which seems to be the case: https://github.com/waveshareteam/e-Paper/tree/master/Arduino/epd4in2b_V2

**Reply by ReiettoAyanami** [reply]
Good morning,
Sorry if i did't text anything yesterday, but here's my updates:

- Updating the constructor to my wires actually made the display works as intented (_Who could've guessed_, right?)
- Bought another E-ink display (only black and white) same brand, same size, just not red anymore. It can display 4 shades of grey, i just don't know how to do that now.
- The constructor for the specified wiring: `GxEPD2_BW<GxEPD2_420_GDEY042T81, GxEPD2_420_GDEY042T81::HEIGHT> display(GxEPD2_420_GDEY042T81(SS, 2, 0, 4));`
Everything is working like a charm, except the wi-fi example, but that's my problem, i guess.

**Reply by ZinggJM** [reply]
Try with https://github.com/ZinggJM/GxEPD2_4G.

**Reply by G6EJD** [reply]
4-levels or more of colour, generally it's down to bit patterns:
#define White         0xFF
#define LightGrey     0xBB
#define Grey          0x88
#define DarkGrey      0x44
#define Black         0x00

Where all bits On 0xFF represents White and all Off 0x00 is black.

This is an 8-bit example, easily extended to 16-bit such as 0xFFFF and so-on

**Reply by ReiettoAyanami** [reply]
One thing, is there already a module/library/piece of code that automatically downloads bitmaps from HTTP?(i dont need https in this case)

**Reply by ZinggJM** [reply]
There is an example that shows how this can be done: 
https://github.com/ZinggJM/GxEPD2/tree/master/examples/GxEPD2_WiFi_Example
for both http: and https:

Don't forget to adapt line 47,48
```
const char* ssid     = "........";
const char* password = "........";
```
to your network.

**Reply by ReiettoAyanami** [reply]
Thanks! I'll take a look into it

**Reply by ReiettoAyanami** [reply]
One last question before I get flamed, is the waveshare spectra e6 supported?

**Reply by ZinggJM** [reply]
I could go searching. But you could as well provide a link!

**Reply by ReiettoAyanami** [reply]
Hi, so sorry, but I forgot, here's the link:
https://www.waveshare.com/4inch-e-paper-hat-plus-e.htm?sku=27367
[Wiki](https://www.waveshare.com/wiki/4inch_e-Paper_HAT%2B_(E)_Manual#Introduction)

I'm pretty sure it's the same [this](https://www.good-display.com/product/532.html) display from Good-Display, but I'm not so confident about it...

EDIT:
I'm also pretty conscious about the fact that it's a full color display and it wouldn't be ideal for an ESP32's memory space(? Correct me if i'm wrong), so i might just ask.

**Reply by ZinggJM** [reply]
No, unfortunately it is not supported. Take a look at either the list in https://github.com/ZinggJM/GxEPD2/blob/master/README.md, or the selectable display drivers in [GxEPD2_display_selection_new_style.h](https://github.com/ZinggJM/GxEPD2/blob/master/examples/GxEPD2_Example/GxEPD2_display_selection_new_style.h).

I also think it is the same display.

The memory space is not an issue. GxEPD2 supports paged display. It uses some trick to cope with the missing partial window addressing.

**Reply by Gauthamraju31** [reply]
@ZinggJM Is there any plan to support the above mentioned GDEP040E01 or waveshare spectra e6?
Thank you.

https://www.good-display.com/product/532.html
https://www.waveshare.com/wiki/4inch_e-Paper_HAT%2B_(E)_Manual#Introduction
